### PR TITLE
Add support to fallback from ENI PD if subnet is out of /28 prefixes

### PR DESCRIPTION
**Pull request by hemanthmalla**
When a subnet is very fragmented, there might not be enough free
contiguous IPs to make up a `/28` block. Without this fallback, new node
creation would be blocked when PD is enabled and subnet is in this
state. Currently operator does not support PD in mixed mode. Once a node
has secondary IP addresses attached on the ENI, operator will not attempt
to allocate any more prefixes.

With this fallback, operator will try to provision new nodes with prefix delegation
enabled. If a subnet doesn't have `/28` blocks available, regular `/32` IPs would
be allocated instead.

AWS VPC CNI plugin uses `v1` of `aws-sdk-go` and has [better error handling](https://github.com/aws/amazon-vpc-cni-k8s/blob/fd8bcf0be4b522d13fb69c18539921452e4dec80/pkg/awsutils/awsutils.go#L1477-L1487).
Out of capacity errors are logged in VPC CNI plugin, but it does not seem to have
any fallback.

### CODE REVIEW

**Review by gandro — APPROVED** [review]
Thanks!

File: pkg/aws/eni/node.go [hunk 1]
```diff
@@ -270,7 +289,15 @@ func (n *Node) AllocateIPs(ctx context.Context, a *ipam.AllocationAction) error
...
-		return n.manager.api.AssignENIPrefixes(ctx, a.InterfaceID, int32(numPrefixes))
+		err := n.manager.api.AssignENIPrefixes(ctx, a.InterfaceID, int32(numPrefixes))
+		if !isSubnetAtCapacity(err) {
+			return err
+		}
+		// Subnet might be out of available /28 prefixes, but /32 IP addresses might be available.
+		// We should attempt to allocate /32 IPs.
+		n.loggerLocked().WithFields(logrus.Fields{
+			logfields.Node: n.k8sObj.Name,
+		}).Warning("Subnet might be out of prefixes, operator will not allocate prefixes on this node anymore")
```
Comment (christarazi) [inline pkg/aws/eni/node.go]:
```suggestion
		}).Warning("Subnet might be out of prefixes, Cilium will not allocate prefixes on this node anymore")
```

**Review by christarazi — APPROVED** [review]

File: pkg/aws/eni/node.go (hunk 1 above)
Comment (hemanthmalla) [reply pkg/aws/eni/node.go]: Thanks. Updated this.

**Review by gandro — APPROVED** [review]

**Review by christarazi — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
@hemanthmalla Could you rebase the PR? I believe some of the build failures were already fixed in master branch.

**Comment by hemanthmalla** [comment]
@aanm build seems to have failed again ?
https://app.travis-ci.com/github/cilium/cilium/jobs/581598034

**Comment by christarazi** [comment]
Yep looks like a known failure: https://github.com/cilium/cilium/issues/21145

**Comment by gandro** [comment]
> K8sConformance Portmap Chaining Check connectivity-check compliance with portmap chaining

Has already been removed in master (#21130) :thinking: Why is it still executed?

**Comment by hemanthmalla** [comment]
Rebasing again to include changes from https://github.com/cilium/cilium/pull/21220

**Pull request merged**